docs/references-docs-in-vfs.md

```diff
@@ -0,0 +1,301 @@
+for each file individually. This route allows to make it in bulk.
+
+#### Request
+
+```http
+POST /files/io.cozy.playlists/e9308dc2-e2e3-11e6-b685-fb88662613d4/relationships/references HTTP/1.1
+Content-Type: application/vnd.api+json
+Accept: application/vnd.api+json
+```
+
+```json
+{
+  "data": [
+    { "type": "io.cozy.files", "id": "417c4e58-e2e4-11e6-b7dc-2b68ed7b77f4" },
+    { "type": "io.cozy.files", "id": "4504f55c-e2e4-11e6-88f2-d77aeecab549" },
+    { "type": "io.cozy.files", "id": "4587727a-e2e4-11e6-bfe9-ef1be7df7f26" },
+    { "type": "io.cozy.files", "id": "45d591d0-e2e4-11e6-ab9a-ff3b218e31cc" }
+  ]
+}
+```
+
+#### Response
+
+```http
+HTTP/1.1 204 No Content
+Content-Type: application/vnd.api+json
+```
+
+### DELETE /files/:type/:doc-id/relationships/references
+
+This bulk deletion of refences on many files can be useful when an album or
+playlist is deleted.
+
+#### Request
+
+```http
+DELETE /files/io.cozy.playlists/e9308dc2-e2e3-11e6-b685-fb88662613d4/relationships/references HTTP/1.1
+Content-Type: application/vnd.api+json
+Accept: application/vnd.api+json
+```
+
+```json
+{
+  "data": [
+    { "type": "io.cozy.files", "id": "417c4e58-e2e4-11e6-b7dc-2b68ed7b77f4" },
+    { "type": "io.cozy.files", "id": "4504f55c-e2e4-11e6-88f2-d77aeecab549" },
+    { "type": "io.cozy.files", "id": "4587727a-e2e4-11e6-bfe9-ef1be7df7f26" },
+    { "type": "io.cozy.files", "id": "45d591d0-e2e4-11e6-ab9a-ff3b218e31cc" }
+  ]
+}
+```
+
+#### Response
+
+```http
+HTTP/1.1 204 No Content
+Content-Type: application/vnd.api+json
+```
+
+## Usage
+
+### Modification of a referenced file
+
+Before an application updates a file, it can check if the file has some
+references. If it is the case, it may offer to the user two choices:
+
+- update the file with the new version (the albums and playlists will use the
+  new version)
+- save the new version as a new file and preserve the old file (the old file
+  may be moved to a `originals` directory).
+
+### Moving a referenced file to the trash
+
+The stack will refuse to move a referenced file to the trash. It's the same
+for a folder that contains a referenced file. The application has to remove
+the references first. It's possible to clear all the references on a file with
+a `PATCH` request like this:
+
+```http
+PATCH /files/9152d568-7e7c-11e6-a377-37cbfb190b4b/relationships/references HTTP/1.1
+Content-Type: application/vnd.api+json
+Accept: application/vnd.api+json
+```
+
+```json
+{
+  "data": []
+}
+```
+
+## Implementation
+
+The references are persisted in the `io.cozy.files` documents in CouchDB. A
+mango index is used to fetch all the files that are associated to a given
+document (for `GET /files/:type/:doc-id/relationships/references`).
+
+For request to update or move to trash a file, it is easy to fetch its
+CouchDB document to see if it has a reference. But it is more difficult when
+moving a folder to trash. To do that, we will add on each folder a counter of
```

Comment:
For option A, I think we will have to make the GETs even without the references, just to be sure that we didn't create a loop (the parent of folder A is B and the parent of folder B is A).

For the PUTs, we can use a local document in couchdb instead of putting this information in the main documents (IIRC, documents those _id starts with an `_` are not replicated in couchdb).